File: Realization_GO/dynamic_array.go
package main

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// DynamicArray - динамический массив строк
type DynamicArray struct {
	data     []string
	size     int
	capacity int
}

// NewDynamicArray - конструктор
func NewDynamicArray(initialCapacity int) *DynamicArray {
	if initialCapacity < 1 {
		initialCapacity = 10
	}
	return &DynamicArray{
		data:     make([]string, initialCapacity),
		size:     0,
		capacity: initialCapacity,
	}
}

// resize - изменение размера массива
func (d *DynamicArray) resize(newCapacity int) {
	newData := make([]string, newCapacity)
	for i := 0; i < d.size; i++ {
		newData[i] = d.data[i]
	}
	d.data = newData
	d.capacity = newCapacity
}

// PushBack - добавление в конец
func (d *DynamicArray) PushBack(value string) {
	if d.size >= d.capacity {
		newCap := d.capacity * 2
		if newCap == 0 {
			newCap = 4
		}
		d.resize(newCap)
	}
	d.data[d.size] = value
	d.size++
}

// Insert - вставка по индексу
func (d *DynamicArray) Insert(index int, value string) {
	if index < 0 || index > d.size {
		fmt.Println("Неверный индекс!")
		return
	}

	if d.size >= d.capacity {
		d.resize(d.capacity * 2)
	}

	for i := d.size; i > index; i-- {
		d.data[i] = d.data[i-1]
	}

	d.data[index] = value
	d.size++
}

// Remove - удаление по индексу
func (d *DynamicArray) Remove(index int) {
	if index < 0 || index >= d.size {
		fmt.Println("Неверный индекс")
		return
	}

	for i := index; i < d.size-1; i++ {
		d.data[i] = d.data[i+1]
	}
	d.size--

	if d.size < d.capacity/4 && d.capacity > 10 {
		d.resize(d.capacity / 2)
	}
}

// Get - получение элемента по индексу
func (d *DynamicArray) Get(index int) string {
	if index < 0 || index >= d.size {
		fmt.Println("Неверный индекс")
		return ""
	}
	return d.data[index]
}

// Set - установка значения по индексу
func (d *DynamicArray) Set(index int, value string) {
	if index < 0 || index >= d.size {
		fmt.Println("Неверный индекс")
		return
	}
	d.data[index] = value
}

// GetSize - получение размера
func (d *DynamicArray) GetSize() int {
	return d.size
}

// GetCapacity - получение емкости
func (d *DynamicArray) GetCapacity() int {
	return d.capacity
}

// Print - печать массива
func (d *DynamicArray) Print() {
	fmt.Printf("Массив[%d/%d]: ", d.size, d.capacity)
	for i := 0; i < d.size; i++ {
		fmt.Printf("\"%s\" ", d.data[i])
	}
	fmt.Println()
}

// Clear - очистка массива
func (d *DynamicArray) Clear() {
	d.size = 0
}

// SaveToFile - сохранение в текстовый файл
func (d *DynamicArray) SaveToFile(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Printf("Ошибка создания файла: %s\n", filename)
		return err
	}
	defer file.Close()

	_, err = fmt.Fprintf(file, "%d\n", d.size)
	if err != nil {
		return err
	}

	for i := 0; i < d.size; i++ {
		_, err = fmt.Fprintf(file, "%s\n", d.data[i])
		if err != nil {
			return err
		}
	}

	fmt.Printf("Массив сохранён в файл: %s\n", filename)
	return nil
}

// LoadFromFile - загрузка из текстового файла
func (d *DynamicArray) LoadFromFile(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия файла: %s\n", filename)
		return err
	}
	defer file.Close()

	d.Clear()

	scanner := bufio.NewScanner(file)

	// Читаем размер
	if !scanner.Scan() {
		return fmt.Errorf("файл пуст")
	}
	n, err := strconv.Atoi(scanner.Text())
	if err != nil {
		return err
	}

	// Читаем строки
	for i := 0; i < n && scanner.Scan(); i++ {
		value := scanner.Text()
		value = strings.TrimRight(value, "\n\r")
		d.PushBack(value)
	}

	fmt.Printf("Массив загружен из файла: %s\n", filename)
	return scanner.Err()
}

// SaveToBinaryFile - сохранение в бинарный файл
func (d *DynamicArray) SaveToBinaryFile(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Printf("Ошибка создания бинарного файла: %s\n", filename)
		return err
	}
	defer file.Close()

	// Запись размера
	err = binary.Write(file, binary.LittleEndian, int32(d.size))
	if err != nil {
		return err
	}

	// Запись строк
	for i := 0; i < d.size; i++ {
		str := d.data[i]
		strLen := int32(len(str))

		// Запись длины строки
		err = binary.Write(file, binary.LittleEndian, strLen)
		if err != nil {
			return err
		}

		// Запись строки
		_, err = file.Write([]byte(str))
		if err != nil {
			return err
		}
	}

	fmt.Printf("Динамический массив сохранён в бинарный файл: %s\n", filename)
	return nil
}

// LoadFromBinaryFile - загрузка из бинарного файла
func (d *DynamicArray) LoadFromBinaryFile(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия бинарного файла: %s\n", filename)
		return err
	}
	defer file.Close()

	d.Clear()

	var fileSize int32
	err = binary.Read(file, binary.LittleEndian, &fileSize)
	if err != nil {
		return err
	}

	for i := 0; i < int(fileSize); i++ {
		var strLen int32
		err = binary.Read(file, binary.LittleEndian, &strLen)
		if err != nil {
			return err
		}

		buffer := make([]byte, strLen)
		_, err = file.Read(buffer)
		if err != nil {
			return err
		}

		d.PushBack(string(buffer))
	}

	fmt.Printf("Динамический массив загружен из бинарного файла: %s\n", filename)
	return nil
}

File: LR3/Realization_C++/DynamicArray.cpp
#include "DynamicArray.h"
#include <iostream>
#include <fstream>
#include <string>

using namespace std;

void DynamicArray::resize(int newCapacity) {
    string* newData = new string[newCapacity];
    for(int i = 0; i < size; i++) {
        newData[i] = data[i];
    }
    delete[] data;
    data = newData;
    capacity = newCapacity;
}

DynamicArray::DynamicArray(int initialCapacity) {
    capacity = initialCapacity;
    size = 0;
    data = new string[capacity];
}

DynamicArray::~DynamicArray() {
    delete[] data;
}

void DynamicArray::pushBack(const string& value) {
    if(size >= capacity) {
        resize(capacity * 2);
    }
    data[size++] = value;
}

void DynamicArray::insert(int index, const string& value) {
    if(index < 0 || index > size) {
        cout << "Неверный индекс!" << endl;
        return;
    }
    if(size >= capacity) {
        resize(capacity * 2);
    }
    for(int i = size; i > index; i--) {
        data[i] = data[i - 1];
    }
    data[index] = value;
    size++;
}

void DynamicArray::remove(int index) {
    if(index < 0 || index >= size) {
        cout << "Неверный индекс" << endl;
        return;
    }
    for(int i = index; i < size-1; i++) {
        data[i] = data[i + 1];
    }
    size--;
    if(size < capacity / 4 && capacity > 10) {
        resize(capacity / 2);
    }
}

string DynamicArray::get(int index) {
    if(index < 0 || index >= size){
        cout << "Неверный индекс" << endl;
        return "";
    }
    return data[index];
}

void DynamicArray::set(int index, const string& value) {
    if(index < 0 || index >= size) {
        cout << "Неверный индекс" << endl;
        return;
    }
    data[index] = value;
}

int DynamicArray::getSize() {
    return size;
}

int DynamicArray::getCapacity() {
    return capacity;
}

void DynamicArray::print() {
    cout << "Массив[" << size << "/" << capacity << "]: ";
    for(int i = 0; i < size; i++) {
        cout << "\"" << data[i] << "\" ";
    }
    cout << endl;
}

void DynamicArray::clear() {
    size = 0;
}

void DynamicArray::saveToFile(const string& filename) {
    ofstream file(filename);
    if(file.is_open()) {
        file << size << endl;
        for(int i = 0; i < size; i++) {
            file << data[i] << endl;
        }
        file.close();
        cout << "Массив сохранён в файл: " << filename << endl;
    } else {
        cout << "Ошибка создания файла: " << filename << endl;
    }
}

void DynamicArray::loadFromFile(const string& filename) {
    ifstream file(filename);
    if(file.is_open()) {
        clear();
        int n;
        file >> n;
        string dummy;
        getline(file, dummy);
        
        for(int i = 0; i < n; i++) {
            string value;
            getline(file, value);
            if (!value.empty() && value.back() == '\n') {
                value.pop_back();
            }
            if (!value.empty() && value.back() == '\r') {
                value.pop_back();
            }
            pushBack(value);
        }
        file.close();
        cout << "Массив загружен из файла: " << filename << endl;
    } else {
        cout << "Ошибка открытия файла: " << filename << endl;
    }
}

// Бинарная сериализация
void DynamicArray::saveToBinaryFile(const std::string& filename) {
    std::ofstream file(filename, std::ios::binary);
    if(file.is_open()) {
        file.write(reinterpret_cast<const char*>(&size), sizeof(int));
        for(int i = 0; i < size; i++) {
            int strLength = data[i].length();
            file.write(reinterpret_cast<const char*>(&strLength), sizeof(int));
            file.write(data[i].c_str(), strLength);
        }
        
        file.close();
        std::cout << "Динамический массив сохранён в бинарный файл: " << filename << std::endl;
    } else {
        std::cout << "Ошибка открытия бинарного файла для записи: " << filename << std::endl;
    }
}

void DynamicArray::loadFromBinaryFile(const std::string& filename) {
    std::ifstream file(filename, std::ios::binary);
    if(file.is_open()) {
        clear();
        
        int fileSize;
        file.read(reinterpret_cast<char*>(&fileSize), sizeof(int));
        
        for(int i = 0; i < fileSize; i++) {
            int strLength;
            file.read(reinterpret_cast<char*>(&strLength), sizeof(int));
            
            char* buffer = new char[strLength + 1];
            file.read(buffer, strLength);
            buffer[strLength] = '\0';
            
            pushBack(std::string(buffer));
            delete[] buffer;
        }
        
        file.close();
        std::cout << "Динамический массив загружен из бинарного файла: " << filename << std::endl;
    } else {
        std::cout << "Ошибка открытия бинарного файла для чтения: " << filename << std::endl;
    }
}